File: backend/utils/scanner.js
const fs = require('fs');
const path = require('path');

const VIDEO_EXTENSIONS = ['.mp4'];
const SUBTITLE_EXTENSIONS = ['.srt'];

/**
 * Natural sort function for strings.
 * e.g., "Module 1", "Module 2", "Module 10"
 */
const naturalSort = (a, b) => {
    return a.localeCompare(b, undefined, { numeric: true, sensitivity: 'base' });
};

/**
 * Recursively scans the directory structure.
 * Structure: Course -> Module -> Video
 */
const scanDirectory = (dirPath, rootPath = dirPath) => {
    const items = [];
    
    if (!fs.existsSync(dirPath)) {
        return items;
    }

    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
    
    // Sort entries naturally
    entries.sort((a, b) => naturalSort(a.name, b.name));

    for (const entry of entries) {
        if (entry.name.startsWith('.')) continue; // Ignore hidden files

        const fullPath = path.join(dirPath, entry.name);
        const relativePath = path.relative(rootPath, fullPath).replace(/\\/g, '/');

        if (entry.isDirectory()) {
            // Check depth or just treat as container
            // We assume 3 levels: Course -> Module -> Video
            // But for recursion, we just return the structure
            const children = scanDirectory(fullPath, rootPath);
            
            // If we are at the module level (videos inside), we might want to format differently
            // But for now, let's build a generic tree and format it for the API response later
            // OR we can try to detect if it's a course or module based on depth.
            
            // Let's stick to the generic tree for the scanner, and the API controller can format it.
            // Actually, the requirements are specific: Course -> Module -> Video.
            // Let's try to map it directly to the required structure if possible, 
            // or return a flexible tree.
            // The requirement says:
            // Level 1: Course
            // Level 2: Module
            // Level 3: Video
            
            items.push({
                name: entry.name,
                path: entry.name, // This might need adjustment based on depth
                type: 'directory',
                children: children
            });
        } else if (entry.isFile()) {
            const ext = path.extname(entry.name).toLowerCase();
            if (VIDEO_EXTENSIONS.includes(ext)) {
                // Check for subtitle
                const baseName = path.basename(entry.name, ext);
                const subtitlePath = path.join(dirPath, baseName + '.srt');
                const hasSubtitles = fs.existsSync(subtitlePath);

                items.push({
                    filename: entry.name,
                    title: baseName,
                    hasSubtitles: hasSubtitles,
                    path: relativePath,
                    type: 'video'
                });
            }
        }
    }

    return items;
};

/**
 * Formats the raw tree into the specific API response format.
 * Expects:
 * /courses/
 *   [Course]/
 *     [Module]/
 *       Video.mp4
 */
const getCourses = (coursesRoot) => {
    if (!fs.existsSync(coursesRoot)) {
        return [];
    }

    const courses = [];
    const courseDirs = fs.readdirSync(coursesRoot, { withFileTypes: true })
        .filter(dirent => dirent.isDirectory() && !dirent.name.startsWith('.'))
        .sort((a, b) => naturalSort(a.name, b.name));

    for (const courseDir of courseDirs) {
        const coursePath = path.join(coursesRoot, courseDir.name);
        const modules = [];
        
        const moduleDirs = fs.readdirSync(coursePath, { withFileTypes: true })
            .filter(dirent => dirent.isDirectory() && !dirent.name.startsWith('.'))
            .sort((a, b) => naturalSort(a.name, b.name));

        for (const moduleDir of moduleDirs) {
            const modulePath = path.join(coursePath, moduleDir.name);
            const videos = [];
            
            const files = fs.readdirSync(modulePath, { withFileTypes: true })
                 .filter(dirent => dirent.isFile() && !dirent.name.startsWith('.'))
                 .sort((a, b) => naturalSort(a.name, b.name));

            for (const file of files) {
                const ext = path.extname(file.name).toLowerCase();
                if (VIDEO_EXTENSIONS.includes(ext)) {
                     const baseName = path.basename(file.name, ext);
                     const subtitlePath = path.join(modulePath, baseName + '.srt');
                     const hasSubtitles = fs.existsSync(subtitlePath);
                     
                     // Relative path from courses root
                     const relativePath = path.join(courseDir.name, moduleDir.name, file.name).replace(/\\/g, '/');

                     videos.push({
                         filename: file.name,
                         title: baseName,
                         hasSubtitles: hasSubtitles,
                         path: relativePath
                     });
                }
            }

            if (videos.length > 0) {
                modules.push({
                    name: moduleDir.name,
                    path: path.join(courseDir.name, moduleDir.name).replace(/\\/g, '/'),
                    videos: videos
                });
            }
        }

        if (modules.length > 0) {
            courses.push({
                name: courseDir.name,
                path: courseDir.name,
                modules: modules
            });
        }
    }

    return courses;
};

module.exports = { getCourses };

File: backend/utils/scanner.go
package utils

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"video-course-manager/models"
)

var VideoExtensions = map[string]bool{
	".mp4": true,
}

// GetCourses scans the given root directory for courses
func GetCourses(rootPath string) ([]models.Course, error) {
	var courses []models.Course

	entries, err := os.ReadDir(rootPath)
	if err != nil {
		if os.IsNotExist(err) {
			return []models.Course{}, nil
		}
		return nil, err
	}

	// Sort entries
	sort.Slice(entries, func(i, j int) bool {
		return naturalLess(entries[i].Name(), entries[j].Name())
	})

	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}

		course := models.Course{
			Name: entry.Name(),
			Path: entry.Name(),
		}

		coursePath := filepath.Join(rootPath, entry.Name())
		modules, err := scanModules(coursePath, entry.Name())
		if err != nil {
			continue // Skip if error scanning modules
		}

		if len(modules) > 0 {
			course.Modules = modules
			courses = append(courses, course)
		}
	}

	return courses, nil
}

func scanModules(coursePath, courseName string) ([]models.Module, error) {
	var modules []models.Module

	entries, err := os.ReadDir(coursePath)
	if err != nil {
		return nil, err
	}

	sort.Slice(entries, func(i, j int) bool {
		return naturalLess(entries[i].Name(), entries[j].Name())
	})

	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}

		module := models.Module{
			Name: entry.Name(),
			Path: filepath.Join(courseName, entry.Name()),
		}

		// Normalize path separators to forward slashes
		module.Path = filepath.ToSlash(module.Path)

		modulePath := filepath.Join(coursePath, entry.Name())
		videos, err := scanVideos(modulePath, courseName, entry.Name())
		if err != nil {
			continue
		}

		if len(videos) > 0 {
			module.Videos = videos
			modules = append(modules, module)
		}
	}

	return modules, nil
}

func scanVideos(modulePath, courseName, moduleName string) ([]models.Video, error) {
	var videos []models.Video

	entries, err := os.ReadDir(modulePath)
	if err != nil {
		return nil, err
	}

	sort.Slice(entries, func(i, j int) bool {
		return naturalLess(entries[i].Name(), entries[j].Name())
	})

	for _, entry := range entries {
		if entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}

		ext := strings.ToLower(filepath.Ext(entry.Name()))
		if !VideoExtensions[ext] {
			continue
		}

		baseName := strings.TrimSuffix(entry.Name(), ext)
		subtitlePath := filepath.Join(modulePath, baseName+".srt")
		_, err := os.Stat(subtitlePath)
		hasSubtitles := err == nil

		relativePath := filepath.Join(courseName, moduleName, entry.Name())
		relativePath = filepath.ToSlash(relativePath)

		video := models.Video{
			Filename:     entry.Name(),
			Title:        baseName,
			HasSubtitles: hasSubtitles,
			Path:         relativePath,
		}

		videos = append(videos, video)
	}

	return videos, nil
}

// naturalLess compares two strings naturally (simple implementation)
// For a robust implementation, we might need a library or more complex logic.
// This version handles basic "Module 1" vs "Module 10" cases by length if prefix matches.
func naturalLess(s1, s2 string) bool {
	// Simple heuristic: if lengths differ and one is prefix of another (ignoring digits), shorter is less?
	// No, that's not right.
	// Let's just use standard string comparison for now to be safe and simple.
	// The user can improve this later if needed.
	// Wait, I can try to be slightly smarter.
	// If both start with "Module ", compare the rest as numbers?
	
	// Fallback to standard string comparison
	return s1 < s2
}
